python/ray/serve/master.py

```diff
@@ -1,334 +1,528 @@
+
+        If the route is None, this is a "headless" endpoint that will not
+        be added to the HTTP proxy (can only be accessed via a handle).
+        """
+        async with self.write_lock:
+            # If this is a headless service with no route, key the endpoint
+            # based on its name.
+            # TODO(edoakes): we should probably just store routes and endpoints
+            # separately.
+            if route is None:
+                route = endpoint
+
+            # TODO(edoakes): move this to client side.
+            err_prefix = "Cannot create endpoint."
+            if route in self.routes:
+                if self.routes[route] == (endpoint, methods):
+                    return
+                else:
+                    raise ValueError(
+                        "{} Route '{}' is already registered.".format(
+                            err_prefix, route))
+
+            if endpoint in self.get_all_endpoints():
+                raise ValueError(
+                    "{} Endpoint '{}' is already registered.".format(
+                        err_prefix, endpoint))
+
+            logger.info(
+                "Registering route {} to endpoint {} with methods {}.".format(
+                    route, endpoint, methods))
+
+            self.routes[route] = (endpoint, methods)
+
+            # NOTE(edoakes): we must write a checkpoint before pushing the
+            # update to avoid inconsistent state if we crash after pushing the
+            # update.
+            self._checkpoint()
+            await self.http_proxy.set_route_table.remote(self.routes)
 
     async def create_backend(self, backend_tag, backend_config, func_or_class,
                              actor_init_args):
-        backend_config_dict = dict(backend_config)
-        backend_worker = create_backend_worker(func_or_class)
-
-        assert backend_tag not in self.get_all_backends(), \
-            "Cannot create backend '{}' because a backend" \
-            "with that name already exists.".format(backend_tag)
-
-        # Save creator which starts replicas.
-        self.backend_table.register_backend(backend_tag, backend_worker)
-
-        # Save information about configurations needed to start the replicas.
-        self.backend_table.register_info(backend_tag, backend_config_dict)
-
-        # Save the initial arguments needed by replicas.
-        self.backend_table.save_init_args(backend_tag, actor_init_args)
-
-        # Set the backend config inside the router
-        # (particularly for max-batch-size).
-        [router] = self.get_router()
-        await router.set_backend_config.remote(backend_tag,
-                                               backend_config_dict)
-
-        await self.scale_replicas(backend_tag,
-                                  backend_config_dict["num_replicas"])
+        """Register a new backend under the specified tag."""
+        async with self.write_lock:
```

Comment:
i'm a little bit concerned about the performance here. we are creating new actors when this lock is held. this is pretty slow, so a concurrent set_traffic_policy call will be delayed by an indefinite amount.

Reply:
Performance for API calls is not what we should optimize for. If users complain about this at some point, we can optimize, but it would be a strange requirement.